src/index.ts

```diff
@@ -0,0 +1,414 @@
+   * This is a pure delegate to the [[modeProperty]].
+   */
+  set mode(value: string) {
+    EditorModel.modeProperty.set(this, value);
+  }
+
+  /**
+   * Get the showLineNumbers flag for the editor model.
+   *
+   * #### Notes
+   * This is a pure delegate to the [[showLineNumbersProperty]].
+   */
+  get showLineNumbers(): boolean {
+    return EditorModel.showLineNumbersProperty.get(this);
+  }
+
+  /**
+   * Set the showLineNumbers flag for the editor model.
+   *
+   * #### Notes
+   * This is a pure delegate to the [[showLineNumbersProperty]].
+   */
+  set showLineNumbers(value: boolean) {
+    EditorModel.showLineNumbersProperty.set(this, value);
+  }
+
+  /**
+   * Get the readOnly flag for the editor model.
+   *
+   * #### Notes
+   * This is a pure delegate to the [[readOnlyProperty]].
+   */
+  get readOnly(): boolean {
+    return EditorModel.readOnlyProperty.get(this);
+  }
+
+  /**
+   * Set the readOnly flag for the editor model.
+   *
+   * #### Notes
+   * This is a pure delegate to the [[readOnlyProperty]].
+   */
+  set readOnly(value: boolean) {
+    EditorModel.readOnlyProperty.set(this, value);
+  }
+
+  /**
+   * Get the tabSize number for the editor model.
+   *
+   * #### Notes
+   * This is a pure delegate to the [[tabSizeProperty]].
+   */
+  get tabSize(): number {
+    return EditorModel.tabSizeProperty.get(this);
+  }
+
+  /**
+   * Set the tabSize number for the editor model.
+   *
+   * #### Notes
+   * This is a pure delegate to the [[tabSizeProperty]]
+   */
+  set tabSize(value: number) {
+    EditorModel.tabSizeProperty.set(this, value);
+  }
+
+  /**
+   * Get the disposed status of this object.
+   */
+  get isDisposed(): boolean {
+    return this._disposed;
+  }
+
+  /**
+   * Updates the buffer stored in the model.
+   *
+   * @param value - A string containing the complete data in the view.
+   */
+  updateBuffer(value: string): void {
+    if (!this.isDisposed) {
+      this._buffer = value;
+    }
+  }
+
+  /**
+   * Save the current buffer in the existing file.
+   */
+  save(): void {
+    // delegate to jupyter-js-services
+    console.log('model save');
+  }
+
+  /**
+   * Rename the current file.
+   */
+  rename(name: string): void {
+    // delegate to jupyter-js-services
+    console.log('model rename');
+  }
+
+  /**
+   * Dispose of this object.
+   */
+  dispose(): void {
+    this._disposed = true;
+  }
+
+  private _disposed = false;
+  private _buffer = '';
+}
+
+
+/**
+ * A widget used for entering text/code.
+ *
+ * #### Notes
+ * This class happens to use Codemirror as the hosted widget for
+ * text entry, but Codemirror is entirely encapsulated
+ * within this class. As long as the IEditorWidget interface is
+ * adhered to, any other text editor widget can be easily
+ * swapped in.
+ */
+export
+class EditorWidget extends Widget {
+
+  /**
+   * Construct an EditorWidget.
+   */
+  constructor(model: IEditorModel) {
+    super();
+    this.addClass('Ph-EditorWidget');
```

Comment:
Can you open up a dialog on Gitter about what CSS prefix we should be using? IIRC the choice was `jp-` but I'm not 100% sure.